doc/sphinx/language/extensions/canonical.rst

```diff
@@ -41,12 +41,57 @@ in :ref:`canonicalstructures`; here only a simple example is given.
+   Then, each time an equation of the form :n:`(x__i _)` |eq_beta_delta_iota_zeta| :n:`c__i` has to be
    solved during the type checking process, :token:`qualid` is used as a solution.
-   Otherwise said, :token:`qualid` is canonically used to extend the field |c_i|
-   into a complete structure built on |c_i|.
+   Otherwise said, :token:`qualid` is canonically used to extend the field :n:`x__i`
+   into a complete structure built on :n:`c__i` when :n:`c__i` unifies with :n:`(x__i _)`.
+
+   The following kinds of terms are supported for the fields :n:`c__i` of :token:`qualid`:
+
+   * Constants (defined through :cmd:`Definition`, :cmd:`Inductive` and constructors thereof,
+     :cmd:`Record` and fields thereof, etc.) and section variables of an active section,
+     applied to zero or more arguments.
+   * Any :token:`sort`.
+   * Literal functions:  `fun … => …`.
+   * Literal, non-dependent function types, i.e. implications: `… -> …`.
+   * Variables bound in :token:`qualid`.
+
+   When searching for a canonical extension, only the head symbol of any existing instance's field :n:`c__i`
+   is considered.
+   We call this head symbol the *key* and we say ":token:`qualid` keys field :n:`x__i` to :n:`k`" when :n:`c__i`'s
+   head symbol is key :n:`k`.
+   Keys are the only piece of information that is used for canonical extension.
+   The keys corresponding to the kinds of terms listed above are:
+
+   * For constants and section variables, potentially applied to arguments:
+     the constant or variable itself, disregarding any arguments.
+   * For sorts: the sort itself.
+   * For literal functions: a disembodied function key denoted `fun _ => _`, disregarding both its domain
+     and body.
+   * For literal functions types: a disembodied implication key denoted `_ -> _`, disregarding both its
+     domain and co-domain.
+   * For variables bound in :token:`qualid`: a catch-all key denoted `_`.
+
+   This means that, for example, `(fun (x : nat) => x+1)` and `(fun (x : bool) => true)` do not constitute
+   distinct keys, and neither do `(some_constant x1)` and `(some_constant (other_constant y1 y2) x2)`.
+
+   Variables bound in :token:`qualid` are considered to match any term for the purpose of canonical
+   extension.
+   This has two major consequences for a field :n:`c__i` keyed to a variable of :token:`qualid`:
+
+   1. Unless another key—and, thus, instance—matches :n:`c__i`, the instance will always be considered by
+      unification.
+   2. :n:`c__i` will be considered overlapping with (i.e. not distinct from) any other canonical instance
+      that keys :n:`x__i` to one of its own respective variables.
+
+   A record's field :n:`x__i` can only be keyed once to each key.
+   Coq will print a warning when :token:`qualid` keys :n:`x__i` to a term
+   whose head symbol is already keyed to by an existing canonical instance.
+   In this case, Coq will not register that particular field :token:`qualid` as a canonical
+   extension.
```

Comment:
I'm not sure "canonical extension" is a standard terminology. Did you see it used elsewhere?

Reply:
The current docs have this phrasing here:
>  Otherwise said, qualid is canonically used to extend the field ci into a complete structure built on ci.

"Canonical extension" seemed in line with that. I'll happily change it, though. We just need a term that is more fine-grained than "canonical instance" since that (I guess) refers to the entire instance whereas my use of "canonical extension" was meant to refer only to one field of an instance.

Reply:
I'd  use "canonical instance" found by indexing a projection and a field value. Of course if "extending a field to the whole record" is already there I cannot complain to you ;-)